fw/cache.c

```diff
@@ -2472,7 +2529,12 @@ __cache_add_node(TDB *db, TfwHttpResp *resp, unsigned long key)
 	T_DBG3("%s: db=[%p] resp=[%p], req=[%p], key='%lu', data_len='%ld'\n",
 	       __func__, db, resp, resp->req, key, data_len);
 
-	/* TODO #522: revalidate existing entries before inserting a new one. */
+	/*
+	 * Remove each entry with current key, even fresh entries. We assume
+	 * that if we at current place, no valid entries are exists.
+	 */
+	tdb_entry_remove(db, key, &tfw_cache_rec_eq_req, resp->req,
```

Comment:
This function should be also called in `tfw_cache_process` instead of `tfw_cache_invalidate` as I think:
```
if (unlikely(req->method == TFW_HTTP_METH_PUT
			     || req->method == TFW_HTTP_METH_DELETE
			     || req->method == TFW_HTTP_METH_POST
			     || tfw_http_req_is_nip(req))
		    && cache_cfg.cache
		    && resp->status >= 200 && resp->status < 400)
			tfw_cache_invalidate(req);
```

Reply:
Fixed